source/extensions/filters/http/local_ratelimit/local_ratelimit.cc

```diff
@@ -0,0 +1,161 @@
+#include "extensions/filters/http/local_ratelimit/local_ratelimit.h"
+
+#include <string>
+#include <vector>
+
+#include "envoy/http/codes.h"
+
+#include "common/http/utility.h"
+
+#include "extensions/filters/http/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace LocalRateLimitFilter {
+
+FilterConfig::FilterConfig(
+    const envoy::extensions::filters::http::local_ratelimit::v3::LocalRateLimit& config,
+    Event::Dispatcher& dispatcher, Stats::Scope& scope, Runtime::Loader& runtime)
+    : status_(toErrorCode(config.status().code())),
+      stats_(generateStats(config.stat_prefix(), scope)),
+      max_tokens_(config.token_bucket().max_tokens()),
+      tokens_per_fill_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.token_bucket(), tokens_per_fill, 1)),
+      fill_interval_(config.has_token_bucket()
+                         ? PROTOBUF_GET_MS_REQUIRED(config.token_bucket(), fill_interval)
+                         : 0),
+      fill_timer_(fill_interval_ > std::chrono::milliseconds(0)
+                      ? dispatcher.createTimer([this] { onFillTimer(); })
+                      : nullptr),
+      runtime_(runtime),
+      filter_enabled_(
+          config.has_filter_enabled()
+              ? absl::optional<Envoy::Runtime::FractionalPercent>(
+                    Envoy::Runtime::FractionalPercent(config.filter_enabled(), runtime_))
+              : absl::nullopt),
+      filter_enforced_(
+          config.has_filter_enabled()
+              ? absl::optional<Envoy::Runtime::FractionalPercent>(
+                    Envoy::Runtime::FractionalPercent(config.filter_enforced(), runtime_))
+              : absl::nullopt),
+      response_headers_parser_(
+          Envoy::Router::HeaderParser::configure(config.response_headers_to_add())) {
+  if (fill_timer_ && fill_interval_ < std::chrono::milliseconds(50)) {
+    throw EnvoyException("local rate limit token bucket fill timer must be >= 50ms");
+  }
+
+  tokens_ = max_tokens_;
+
+  if (fill_timer_) {
+    fill_timer_->enableTimer(fill_interval_);
+  }
+}
+
+FilterConfig::~FilterConfig() {
+  if (fill_timer_ != nullptr) {
+    fill_timer_->disableTimer();
+  }
+}
+
+LocalRateLimitStats FilterConfig::generateStats(const std::string& prefix, Stats::Scope& scope) {
+  const std::string final_prefix = prefix + ".http_local_rate_limit";
+  return {ALL_LOCAL_RATE_LIMIT_STATS(POOL_COUNTER_PREFIX(scope, final_prefix))};
+}
+
+// Snitched from the local rate limit network filter.
+// TODO(rgs1): reuse the code from above.
```

Comment:
Can we do this now? This code is a bit tricky and I think it should be easy to share it in a utility class/function(s)?

Reply:
yeah absolutely.